### PR TITLE
fix(types): validate Filter operator on direct construction and accept factory-method aliases

### PR DESCRIPTION
**Pull request by jaredmixpanel**
## Summary

`Filter(...)` built positionally with the *factory-method* spelling of the operator (`"greater_than"`, `"is_set"`, boolean `"equals"`) serialized that string verbatim as `filterOperator`, and the Mixpanel query API rejected it with HTTP 400 one round trip later (`Unsupported operator for filter_type number: greater_than`). The factory methods were always correct; only direct construction was unvalidated. Bug report: `context/mp-bench-headless/bug-reports/python-filter-direct-construction-unvalidated-operator.md` (implements its option 2).

`Filter.__post_init__` now:

- **Validates** `_operator` against the `FilterOperator` literal and raises `ValueError` immediately. The message names the bad operator, lists the valid wire operators, and points at the factory methods (and lists the accepted aliases).
- **Accepts the factory-method spellings as aliases** and normalizes them to the wire operator, so the positional call produces byte-identical output to the equivalent factory. The alias table is derived from the public classmethods on `Filter` (a test introspects the class and fails if a factory is added without an alias decision): `not_equals`, `not_contains`, `greater_than`, `less_than`, `between`, `not_between`, `at_least`, `at_most`, `is_set`, `is_not_set`, `starts_with`, `ends_with`, `is_true`, `is_false`, `in_cohort`, `not_in_cohort`, `on`, `not_on`, `before`, `since`, `in_the_last`, `not_in_the_last`, `date_between`, `date_not_between`, `in_the_next`. `equals`, `contains`, `list_contains` already are the wire spelling.
- Keeps the two segmentation-`where` spellings that pre-date the literal constructible: `"is equal to"` -> `"equals"` and `"between"` -> `"is between"` (`NUMBER_OPERATOR_MAP` maps both pairs to the same `==` / `><`, so segfilter output is unchanged).
- **Boolean collapse**: on a `boolean` property, `equals` / `does not equal` with a `True` / `False` value (bare or single-element list) collapse to `true` / `false` with `filterValue: null`, matching `Filter.is_true()` / `Filter.is_false()`. Any other operator on a boolean property is rejected.
- **Never rewrites already-valid input**: a wire operator and its value serialize exactly as given.

The `Filter` docstring no longer claims the class is "never instantiated directly"; it now says the factories are the recommended path and direct construction is validated with aliases accepted. The guard is a plain `ValueError` (not a new coded `ParamValidationError`) because the conformance contract pins the coded-guard registry at 126 entries.

## Before / after

| Call | `filterType` | `filterOperator` before | after | `filterValue` after | Result |
|---|---|---|---|---|---|
| `Filter.greater_than("gold", 10)` | number | `is greater than` | `is greater than` | 10 | 200 |
| `Filter("gold", "greater_than", 10, "number")` | number | `greater_than` (400) | `is greater than` | 10 | 200 |
| `Filter.is_set("class")` | string | `is set` | `is set` | null | 200 |
| `Filter("class", "is_set", None)` | string | `is_set` (400) | `is set` | null | 200 |
| `Filter.is_true("flag")` | boolean | `true` | `true` | null | 200 |
| `Filter("flag", "equals", True, "boolean")` | boolean | `equals` (400) | `true` | null | 200 |
| `Filter("gold", "bigger_than", 10, "number")` | — | `bigger_than` (400) | `ValueError` at construction | — | — |

## Conformance

**Expected CI failure.** The `conformance` job fails on this PR in the "Record-mode drift check" step (97 findings: 93 new vectors from the direct-construction tests, two bundle headers, `manifest.json`, and one changed vector, the segfilter `is equal to` case whose recorded `_operator` becomes `equals`). Under the two-step corpus protocol in `conformance/record/README.md` that failure is the expected signal; the vectors are recorded in a follow-up re-pin PR stamped with this change's squash SHA (shared with #235).

`Filter.__post_init__` running on construction had two knock-on effects on the corpus tooling, both resolved with **tooling** changes — nothing under `conformance/vectors/`, `conformance/goldens/`, the manifest, or the stamps is touched:

- A private `mixpanel_headless.types._filter_unchecked(**fields)` rebuilds a `Filter` field-for-field via `object.__new__` + `object.__setattr__`, without running `__post_init__`. It exists for the codec and for tests / the differential harness that drive the downstream builder guards (engage `ES13`, segfilter `SG1`/`SG2`/`SG3`) with operators the constructor now rejects. `tests/conftest.py::make_unchecked_filter` is a positional wrapper over it.
- `conformance/record/codecs.py` now rehydrates `$type: Filter` payloads through `_filter_unchecked` instead of `cls(**kwargs)` (every other type still goes through its constructor). Rationale: the codec must rebuild the *recorded* object faithfully. The pinned vectors capture the builders' own guard behaviour on an already-constructed Filter; re-running the constructor on replay would either refuse to rebuild those inputs (`"was frobnicated"`, `"magical_unicorn"`, ...) or silently rewrite alias spellings before the builder under test saw them. The builders' behaviour on the rebuilt Filter is unchanged, so the pinned outputs still hold.
- `conformance/differential/strategies.py` builds its deliberately invalid ES13 probes (the import-time edges and the `_escaping_filters` draws) through `_filter_unchecked`.

Result: the 11 pinned vectors that rehydrate a Filter with a non-literal operator (engage `filter_to_selector` / `filters_to_selector` ES13 x4, segfilter SG1/SG2/SG3 seams + edge cases x6, segfilter `is equal to` x1) replay clean, and `conformance/tests` (including the fuzz-harness import) is green. `just check` passes end to end.

## Test plan

- [x] `tests/unit/test_query_types.py::TestFilterDirectConstruction`: the report's three rows assert `build_filter_entry(positional) == build_filter_entry(factory)` and `json.dumps(..., sort_keys=True)` byte-equality; boolean collapse table (equals/not-equals x True/False, list-wrapped bools, `is_true`/`is_false` aliases); boolean rejects `is set` / string / int / None / two-element list; unknown operator raises `ValueError` whose message contains the operator, every wire operator, and `Filter.greater_than`; near-miss spellings (`""`, upper-case, trailing space, `>`) rejected; already-valid wire operators pass through untouched; LC1 guard still runs; alias keys are factory names (plus the documented `is equal to`), alias values are literal members, every factory's emitted operator matches the table, and every factory result rebuilt positionally via its own name is byte-identical.
- [x] `tests/unit/test_query_types.py::TestFilterUnchecked`: verbatim fields, no alias normalization, dataclass defaults, equality with normal construction for valid fields, `TypeError` on missing / unknown fields, instance still frozen.
- [x] `tests/unit/test_query_pbt.py::TestFilterOperatorValidationInvariant` (Hypothesis): every `FilterOperator` member accepted unchanged; every alias maps into the literal; every other `st.text()` string raises; case/whitespace variants raise; boolean collapse equals the `is_true`/`is_false` twin; boolean rejects every other wire operator; `is equal to` / `between` yield the same segfilter entry as their literal twins across random numeric values.
- [x] `tests/unit/test_segfilter.py`: `test_is_equal_to_number` back to plain construction (proves the alias), plus `is equal to`≡`equals` and `between`≡`is between` output equality; genuinely invalid operators use `make_unchecked_filter`.
- [x] `conformance/tests/test_codecs_roundtrip.py`: Filter payloads with rejected operators decode verbatim; alias spellings are not normalized on decode; unknown fields still rejected.
- [x] Existing ES13 / SG* / `BookmarkValidationError`-wrapping tests keep their ids and expectations via the bypass helper.
- [x] `just check` green (lint, fmt-check, typecheck, docstring-cov, test-cov, conformance, build); full non-live suite passes; 11 previously affected vectors replay clean.

## Follow-ups

- **Conformance vector family for `Filter.__post_init__`** (positional construction, alias normalization, boolean collapse, the `ValueError` on unknown operators): a follow-up re-pin PR per the corpus stamp protocol, so the TS port stops agreeing with Python only by both being permissive.
- Re-recording the touched tests will change the recorded `_operator` for `test_is_equal_to_number` from `"is equal to"` to `"equals"`; that lands with the same re-pin.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/mixpanel_headless/types.py [hunk 1]
```diff
@@ -7190,21 +7295,78 @@ class Filter:
...
+            )
+        if (
+            self._property_type == "boolean"
+            and operator not in _BOOLEAN_FILTER_OPERATORS
+        ):
+            raise ValueError(
+                f"Filter operator {operator!r} is not valid for a boolean property "
+                f"({self._property!r}); boolean filters only support 'true' and "
+                "'false' with no value. Use Filter.is_true() / Filter.is_false(), "
+                "or pass value True / False with operator 'equals'."
```
[the remark this answers is not in the document]
Comment (jaredmixpanel) [reply src/mixpanel_headless/types.py]: Fixed in 4f75966. `true` / `false` (and the `is_true` / `is_false` aliases) now raise `ValueError` when the value is not `None`, on every property type. Covered by 32 example cases and a Hypothesis test over text, bool, int, and list values.

File: src/mixpanel_headless/types.py [hunk 2]
```diff
@@ -7190,21 +7295,78 @@ class Filter:
...
+            operator = _FILTER_OPERATOR_ALIASES[operator]
+        if self._property_type == "boolean" and operator in (
+            "equals",
+            "does not equal",
+        ):
+            truth = _boolean_filter_value(self._value)
+            if truth is not None:
+                operator = "true" if truth == (operator == "equals") else "false"
+                object.__setattr__(self, "_value", None)
+        if operator not in _FILTER_WIRE_OPERATORS:
```
[the remark this answers is not in the document]
Comment (jaredmixpanel) [reply src/mixpanel_headless/types.py]: Fixed in 4f75966. An `isinstance(operator, str)` guard runs before any set or dict lookup, so a list, dict, `None`, or int raises the same `ValueError` with the same "Unknown Filter operator" message. Examples and a Hypothesis test cover these.

File: tests/conftest.py [hunk 3]
```diff
@@ -337,3 +338,48 @@ def handler(_request: httpx.Request) -> httpx.Response:
         return httpx.Response(429, headers={"Retry-After": "60"})
 
     return handler
+
+
+def make_unchecked_filter(
+    property: str,
+    operator: str,
+    value: Any,
```
[the remark this answers is not in the document]
Comment (jaredmixpanel) [reply tests/conftest.py]: Fixed in 4f75966. The test helper's `value` is now typed with the new `FilterValue` alias (the exact union `Filter._value` stores). The codec keeps `Any` for decoding arbitrary JSON, with a one-line `# Any: <reason>` justification on each use.

File: src/mixpanel_headless/types.py [hunk 4]
```diff
@@ -7133,13 +7144,98 @@ def __post_init__(self) -> None:
...
-    Each class method maps to specific filterType, filterOperator, and
-    filterValue format in the bookmark JSON.
+    The factory classmethods (``Filter.equals``, ``Filter.greater_than``,
+    ``Filter.is_set``, ...) are the recommended way to build a Filter: each
+    one maps to a specific filterType, filterOperator, and filterValue format
+    in the bookmark JSON. Direct construction is supported and validated:
+    ``_operator`` must be a :data:`FilterOperator` member or a factory-method
+    name (``"greater_than"``, ``"is_set"``, ...), which ``__post_init__``
+    normalizes to the wire spelling so both paths serialize identically.
+    Anything else raises ``ValueError`` at construction instead of surfacing
```
[the remark this answers is not in the document]
Comment (jaredmixpanel) [reply src/mixpanel_headless/types.py:7277]: Fixed in 4f75966. `_literal_types.FilterOperatorInput` is the union of `FilterOperator`, the factory alias names, and `"is equal to"`; `_operator` is annotated with it, and `__post_init__` stores the canonical `FilterOperator`. The type-ignores in the tests are removed and `mypy --strict` is clean. A test locks `get_args(FilterOperatorInput)` to the wire set plus the alias table.

File: src/mixpanel_headless/types.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaredmixpanel) [reply src/mixpanel_headless/types.py]: Fixed in 4f75966. Non-string operators are rejected with `ValueError` before the frozenset lookup, so direct construction has one failure mode.
